kore/src/Kore/Step.hs

```diff
@@ -49,119 +57,152 @@ import qualified Kore.Step.Simplification.Pattern as Pattern
     ( simplifyTopConfiguration
     )
 import Kore.Step.Simplification.Simplify as Simplifier
+
+import Data.Limit
+    ( Limit (..)
+    )
+import qualified Data.Limit as Limit
+import Kore.Debug
 import qualified Kore.Step.SMT.Evaluator as SMT.Evaluator
     ( filterMultiOr
     )
-import qualified Kore.Step.Step as Step
 import Kore.Step.Strategy hiding
     ( transitionRule
     )
 import qualified Kore.Step.Strategy as Strategy
-import qualified Kore.Step.Transition as Transition
 import qualified Kore.Unification.Procedure as Unification
 
-{- | A strategy primitive: a rewrite rule or builtin simplification step.
- -}
-data Prim rewrite = Simplify | Rewrite !rewrite
+{- | The program's state during symbolic execution.
+-}
+data ProgramState a = Start !a | Rewritten !a | Remaining !a
+    deriving (Eq, Ord, Show)
+    deriving (Functor)
+    deriving (GHC.Generic)
+    deriving anyclass (SOP.Generic, SOP.HasDatatypeInfo)
+    deriving anyclass (Debug, Diff)
+
+extractProgramState :: ProgramState a -> a
+extractProgramState (Rewritten a) = a
+extractProgramState (Remaining a) = a
+extractProgramState (Start a) = a
+
+retractRemaining :: ProgramState a -> Maybe a
+retractRemaining (Remaining a) = Just a
+retractRemaining _ = Nothing
+
+{- | The sequence of transitions for the symbolic execution strategy.
+-}
+executionStrategy :: Stream (Strategy Prim)
+executionStrategy =
+    (Strategy.sequence . fmap Strategy.apply)
+        [ Begin
+        , Simplify
+        , ApplyRewrites
+        ]
+    & Stream.iterate id
+
+limitedExecutionStrategy :: Limit Natural -> [Strategy Prim]
+limitedExecutionStrategy depthLimit =
+    Limit.takeWithin depthLimit (toList executionStrategy)
+
+{- The primitive transitions for the symbolic execution strategy.
+-}
+data Prim
+    = Begin
+    | Simplify
+    | ApplyRewrites
```

Comment:
Unless this name is already taken, I suggest changing to a simple imperative verb to match the other constructors:

```suggestion
    | Rewrite
```